Post: The best test is the test you don't have to write (#6)
guillaumewuip/guillaumewuip.github.io: _posts/2021-02-12-the-best-test-is-the-test-you-dont-have-to-write.md

---
title: The best test is the test you don't have to write

draft: false

page : false

image: 

excerpt: ""

---

When a software engineering team wants to improve or secure its codebase

> quality, it often invests in tests heavily. 

Reviewer: ```suggestion
quality, it often heavily invests in tests. 
```

But tests are not a magic remedy. They're not free.

Using tests like the *only* tool that can help with quality can hide the real
questions. Why are we writting tests? How to maintain or improve the quality of

> a codebase without adding too much tests?

Reviewer: ```suggestion
Using tests as the *only* tool that can help with quality can lead to more troubles. Are they all so helpful? Are they worth the cost of their conception and maintenance? How can we maintain or improve the quality of
a codebase without bloating it with superficial tests?
```

## About tests

We write some code and boom! it's time to add the corresponding test. We're
mature now so we know we need different levels of testing to cover a feature:

> unit testing, integration testing, end-to-end testing, snapshot testsing, etc.

Reviewer: ```suggestion
Let's say we wrote some code. Disciplined as we are, we know It's time to add the corresponding test. We have
the maturity to know we need different levels of testing to cover a feature:
unit testing, integration testing, end-to-end testing, snapshot testing, etc.
```

Continuously adding tests create issues on its own. All the tests become a
"codebase in the codebase": with the dedicated libraries we use for them, the
testing framework that is so hard to change, the CI time required to run the
whole test suite, the maintenance tests need, their flakiness, the colleagues 
complaining they need the best in class computer to run them in an acceptable

> time, etc.

Reviewer: ```suggestion
It's a common reflex and our intentions are good. However, continuously adding tests does create issues on its own. All the tests become a
"codebase in the codebase": with the dedicated libraries we use for them, the
testing framework that is so hard to change, the CI time required to run the
whole test suite, the maintenance those tests need, their flakiness, the colleagues 
complaining they need the best in class computer to run them in an acceptable
time, etc.
```

> Tests are not free. Why are we writing them in the first place?

Reviewer: ```suggestion
Tests are not free, so let's challenge our good intentions. Why are we writing them in the first place?
```

We write tests to assert our code perform correctly in certain conditions.

> Because writing code is not easy, a lot can go wrong.

Reviewer: ```suggestion
We write tests to assert our code performs correctly under given constraints.
We know writing code is not easy, and a lot can go (and has gone) wrong.
```

## Being its own enemy

The key element in the last sentence is "in certain conditions". Would the world
always be in one single state we won't had to write and maintain any test: our
code would run the way we've tested it by hand the first time.

> 

Reviewer: ```suggestion
The key element is that tests are performed under "given contraints". If the world
always was in one single and predictable state, we would not have to write and maintain any test: our
code would run the way we've tested it by hand the first time.

```

But the world can be in an infinite number of states and our code need to

> accomodate a certain (big) number of them.

Reviewer: ```suggestion
But the world can be in an infinite number of states. To be always safe, our code and therefore our tests would need to
accommodate as many of them as possible.
```

It's especially true when we write code that will be used by other code. The
larger the API, the more state we allow and the more tests we will have to
write.

The challenge here is for the code to cover all the use cases it needs to cover,
but not more. Everything the code *can* do but *should not* do when used will
require a test asserting it effectively isn't doing that.

> In other words: we should try to not write code that allow *impossible states*. 

Reviewer: ```suggestion
In other words we should not write code that allows *impossible states*. 
```

## Impossible state

> Impossible states are any states our code should not be in. A password that is
> empty during a login. A payment that is at the same time waiting for approval
> and rejected. A request that is both in loading state and success state.

Reviewer: ```suggestion
Impossible states are any states our code should not be in. A password that is
empty during a login. A payment that is at the same time waiting for approval
and rejected. A request that is both loading and successful.
```

> Let's take the request example by writing a naive React hook that does some kind
> of a request and returns data. One can use it like so:

Reviewer: ```suggestion
Let's take the request example by writing a naive React hook that does an unspecified request and returns data. One can use it like so:
```

```typescript
const {
  isLoading, // true | false 
  error,     // Error | undefined
  data,      // TheData | undefined
} = useMyRequest()
```

> How many states do we have here? 2\*2\*2 = 8 states. How many states a naive

Reviewer: ```suggestion
How many states do we have here? 2\*2\*2 = 8 states. How many states a naive
```
Ces 8 states ne sont pas évidents. Tes commentaires en indiquent 6

Author: Ah non !

| isLoading | error | data |
|-----------|-----|------|
| true | undefined | undefined |
| true | undefined | Data |
| true | Error | undefined |
| true | Error | Data |
| false | undefined | undefined |
| false | undefined | Data |
| false | Error | undefined |
| false | Error | Data |

Author: Peut-être que je dois les préciser du coup ?

Reviewer: Yes je crois!

request is in? Let say, 4: request not asked, request loading, request failure,
request success. That means we have to deals with 4 impossible states like this
one, at the same time in loading and with an error:

```typescript
{
  isLoading: true,
  error: Error('something went wrong'),
  data: undefined,
}
```

To assert thoses 4 impossible states never append despite the API allowing them,
we need tests.

Is it irremediable? It should not be. There is enough valid states to handle
already to not add impossible ones.

So, how to [make impossible states

> impossible](https://www.youtube.com/watch?v=IcgmSRJHu_8)?

Reviewer: J'ai de la peine à encaisser cet section haha, je la trouve difficilement lisible...

Reviewer: Dans l'idéal: 
- exemple de l'usage 
- Expliquer comment dans l'usage ça peut créer des impossible states
- Expliquer les 8 tests qui en découlent

Ensuite tu peux dérouler car tu as démontrer le coût de ces impossible states

## Representing knowledge in code

> [There is more to the software engineering job than producing code (and tests)
> continuously](http://guillaume.wuips.com/tech/posts/2020-12-01-defining-success-engineering-team).
> We sometimes have to pause and reflect on what models we're using, what business
> feature we're implementing and use code to write this knowledge down.

Reviewer: ```suggestion
[There is more to software engineering than continuously producing code (and tests)](http://guillaume.wuips.com/tech/posts/2020-12-01-defining-success-engineering-team).
We sometimes have to pause and reflect on what models we're using, what business
features we're implementing. We can then use code to write this knowledge down.
```

Now that we know our request can be in 4 possibles states, how does it look like
in code?

Such a mindset is something we're more used to see in codebases written with
ML-languages like Haskell:

```haskell
data Request = NotAsked
             | Loading
             | Failure Error
             | Success TheData
```

> But nothing prevent us to write something similar with more classic languages,
> here Typescript: 

Reviewer: ```suggestion
But nothing prevents us from writing something similar with more familiar languages,
here in Typescript: 
```

```typescript
type Request<E, P> = 
  | { _type: 'NotAsked' }
  | { _type: 'Loading' }
  | { _type: 'Failure', error: E } 
  | { _type: 'Success', payload: P }
```

> We can clearly see the benefit here. The request can't be at the same time in
> loading and with an error. 

Reviewer: ```suggestion
This typing explicits our targeted benefits. The request can't be at the same time in
loading and with an error. 
```

> Working with such model is both easy and secure: we're forced to handle each
> state to access the one with data but we're guided by the 4 states defined
> earlier:

Reviewer: ```suggestion
Working with such model is both easy and secure: we're forced to handle each
state in order to eventually access the one with data and we're guided by the 4 states defined
earlier:
```

```tsx
function Component() {
  const request = useMyRequest()

  switch (request._type) {
    case 'NotAsked':
      return <NotAskedScreen />

    case 'Loading':
      return <LoadingScreen />

    case 'Failure':
      return <ErrorScreen error={request.error} />

    case 'Success':
      return <SuccessScreen data={request.payload} />
  }
}
```

> We have covered all valid states and no impossible ones. The developper writing
> the `useMyRequest` hook has 4 less tests to write. The one using the hook, being
> unsure of the quality of the work of his colleague, has also 4 tests less to
> write. Everyone is happy and can come home early. Good work!

Reviewer: ```suggestion
We have covered all valid states and no impossible ones. The developper writing
the `useMyRequest` hook has 4 less tests to write: the typing of the hook itself enforces the output can ever only be these four cases. Testing it further would be testing the language itself. Developers using the hook also have 4 less tests to write as a result. Everyone is happy and can go home early. Good work!
```

## The tools

> They are tools that can help us writing such kind of code.

Reviewer: ```suggestion
There are tools available which help write code with such a mindset.
```

> Typed language, first. They are not mandatory but they certainly can help a
> developper write and use such code more easily.

Reviewer: ```suggestion
Using a typed language should be the priority. Though they are not mandatory, they certainly help
developers avoid blind spots in their code. It's the low-hanging fruit of code quality.
```

> Mathematics, with the type theory and especially [Alebraic Data
> Type](https://en.wikipedia.org/wiki/Algebraic_data_type) (ADT). The `Request`
> type we written just above is an ADT. 

Reviewer: ```suggestion
Mathematics also help with type theory and especially [Alebraic Data
Type](https://en.wikipedia.org/wiki/Algebraic_data_type) (ADT). The `Request`
type we wrote above is an ADT. 
```

> Libraries that leverage thoses concepts and functional programing.  Don't bother
> writing the `Request` by hand.  [`RemoteData` is a
> library](https://github.com/devexperts/remote-data-ts) that does just that.
> [`fp-ts` also offer a lot of ADTs](https://github.com/gcanti/fp-ts) to model
> what's going on in our codebase.

Reviewer: ```suggestion
Libraries that leverage those concepts and functional programming are strong allies as well. 
Don't bother writing the `Request` by hand.  [`RemoteData` is a
library](https://github.com/devexperts/remote-data-ts) that does just that.
[`fp-ts` also offer a lot of ADTs](https://github.com/gcanti/fp-ts) to model
what's going on in our codebase. The thing is, our applications are not that different. If you are faced with a feature and ponders on its modelling as code, chances are a tool has been made to help you do just that.
```

> Last but not least, our brains! It's not just about using technical and
> low-level concepts such as `Request`, `Option`, `Task`. Is also about using the
> same mindset to reprent *business* states the codebase should be in.

Reviewer: ```suggestion
Last but not least, our brains! It's not just about using technical and
low-level concepts such as `Request`, `Option`, `Task`. Is also about using the
same mindset to represent *business* as states the codebase should be in.
```

## Conclusion 

Does it mean we should stop writing tests? Certainly not. Tests are absolutly
necessary to prevent our codebase to fail.

> But remember, we need tests because our code can fail. Reducing the number of
> way our code can fail should be our priority as much as writing tests for when
> it can.

Reviewer: ```suggestion
But remember, we need tests because our code can fail. Reducing the number of
way our code can fail should be our priority. We can then keep writing tests for the few remaining chances of failure.
```

---
title: The best test is the test you don't have to write

draft: false

page : false

image: 

excerpt: ""

---

When a software engineering team wants to improve or secure its codebase
quality, it often heavily invests in tests. 

But tests are not a magic remedy. They're not free.

Using tests as the *only* tool that can help with quality can lead to more
troubles. Are they worth the cost of their conception and maintenance? How can
we maintain or improve the quality of a codebase without bloating it with
superficial tests?

## About tests

> Let's say we wrote some code. Disciplined as we are, we know It's time to add
> the corresponding test. We have the maturity to know we need different levels of
> testing to cover a feature: unit testing, integration testing, end-to-end
> testing, snapshot testing, etc.

Reviewer: ```suggestion
Let's say we wrote some code. Disciplined as we are, we know it's time to add
the corresponding test. We have the maturity to know we need different levels of
testing to cover a feature: unit testing, integration testing, end-to-end
testing, snapshot testing, etc.
```

codebase": with the dedicated libraries we use for them, the testing framework
that is so hard to change, the CI time required to run the whole test suite, the
maintenance those tests need, their flakiness, the colleagues complaining they
need the best in class computer to run them in an acceptable time, etc.

Tests are not free, so let's challenge our good intentions. Why are we writing
them in the first place?

We write tests to assert our code performs correctly under given constraints.
We know writing code is not easy, and a lot can go (and has gone) wrong.

## Being its own enemy

The key element is that tests are performed under "given contraints". If the
world always was in one single and predictable state, we would not have to write
and maintain any test: our code would run the way we've tested it by hand the
first time.

But the world can be in an infinite number of states. To be always safe, our
code and therefore our tests would need to accommodate as many of them as
possible.

It's especially true when we write code that will be used by other code. The
larger the API, the more states we allow and the more tests we will have to

> write. We're sometimes our own enemy.

Reviewer: ```suggestion
write. Hell is paved with good intentions.
```

require a test asserting it effectively isn't doing that.

In other words we should not write code that allows *impossible states*. 

## Impossible state

Impossible states are any states our code should not be in. A password that is
empty during a login. A payment that is at the same time waiting for approval
and rejected. A data we're fetching that is both loading and successful.

Let's take the example of the data we're fetching, or *remote data*, by writing
a naive React hook that does an unspecified request and returns data. One can
use it like so:

```typescript
const {
  isLoading, // true | false 
  error,     // Error | undefined
  data,      // TheData | undefined
} = useMyRequest()
```

How many states can be represented by the API here? 2\*2\*2 = 8 states. Not
conviced? Here is the list of all of them:

<figure>
  <div markdown="1">

| isLoading | error | data |
|-----------|-----|------|
| true | undefined | undefined |
| true | undefined | Data |
| true | Error | undefined |
| true | Error | Data |
| false | undefined | undefined |
| false | undefined | Data |
| false | Error | undefined |
| false | Error | Data |

  </div>
  <figcaption>

>     All the states we can represent

Reviewer: ```suggestion
Each line is a possible state created by this hook
```

  </figcaption>
</figure>

> How many states a naive request is in? Let say, 4: request not yet asked,
> request loading, request failure, request success.

Reviewer: ```suggestion
That's  a lot of possibilities and some of them are clearly not needed. After all, how many actual states can or should a naive request be in? Probably only 4: request not yet asked,
request loading, request failure, request success.
```

> That means the API we chose leaves us with 4 valid and 4 impossible states:

Reviewer: ```suggestion
The API we chose with the hook above therefore leaves us with 4 valid and 4 impossible states:
```

<figure>
  <div markdown="1">

| isLoading | error | data |
|-----------|-----|------|
| true | undefined | undefined |
| false | undefined | undefined |
| false | undefined | Data |
| false | Error | undefined |

  </div>
  <figcaption>
  The valid states
  </figcaption>
</figure>

> `false | Error | undefined` for example means "request completed, but resulted
> in failure". Valid state.

Reviewer: ```suggestion
`false | Error | undefined` for example means "request completed, but resulted
in failure", which is one of the valid states.
```

<figure>
  <div markdown="1">

| isLoading | error | data |
|-----------|-----|------|
| true | undefined | Data |
| true | Error | undefined |
| true | Error | Data |
| false | Error | Data |

  </div>
  <figcaption>
  The impossible states
  </figcaption>
</figure>

> `false | Error | Data` means nothing for our naive request (both error and
> data). Impossible state.

Reviewer: ```suggestion
`true | Error | Data` means nothing for our naive request (both error and
data while still loading). It's an impossible state unrepresentative of the actual, actionable code functionality.
```

> To assert thoses 4 impossible states never append despite the API allowing them,
> we need tests.

Reviewer: ```suggestion
To assert those 4 impossible states never happen despite the API allowing them,
we would need to write tests around the API, like an alarm system. "The things you knew could happen but did not want to happen...they did happen".
```

> Is it irremediable? It should not be. There is enough valid states to handle
> already to not add impossible ones.

Reviewer: ```suggestion
Is it irremediable? It should not be. Testing the four valid states is enough work
already, why deal with the impossible ones?
```

So, how to [make impossible states
impossible](https://www.youtube.com/watch?v=IcgmSRJHu_8)?

## Representing knowledge in code

[There is more to software engineering than continuously producing code (and
tests)](http://guillaume.wuips.com/tech/posts/2020-12-01-defining-success-engineering-team).
We sometimes have to pause and reflect on what models we're using, what business
features we're implementing. We can then use code to write this knowledge down.

Now that we know our request can be in only 4 possibles states, how does it look
like in code?

Such a mindset is something we're more used to see in codebases written with
ML-languages like Haskell:

```haskell
data RemoteData = NotAsked
                | Loading
                | Failure Error
                | Success TheData
```

> But nothing prevents us from writing something similar with more familiar
> languages, here in Typescript: 

Reviewer: ```suggestion
But we can also write something similar using more familiar
languages, here in Typescript: 
```

```typescript
type RemoteData<E, P> = 
  | { type: 'NotAsked' }
  | { type: 'Loading' }
  | { type: 'Failure', error: E } 
  | { type: 'Success', payload: P }
```

This typing explicits our targeted benefits. The request can't be at the same
time in loading and with an error. 

> Working with such model is both easy and secure: we're forced to handle each
> state in order to eventually access the one with data and we're guided by the 4
> states defined earlier:

Reviewer: ```suggestion
Working with such a model is both easy and secure: we're forced to handle each
state in order to eventually access the one with data, and we're limited to the 4
states defined earlier:
```

```tsx
function Component() {
  const request = useMyRequest()

  switch (request._type) {
    case 'NotAsked':
      return <NotAskedScreen />

    case 'Loading':
      return <LoadingScreen />

    case 'Failure':
      return <ErrorScreen error={request.error} />

    case 'Success':
      return <SuccessScreen data={request.payload} />
  }
}
```

We have covered all valid states and no impossible ones. The developper writing
the `useMyRequest` hook has 4 less tests to write: the typing of the hook itself
enforces the output can ever only be these four cases. Testing it further would
be testing the language itself. Developers using the hook also have 4 less tests
to write as a result. Everyone is happy and can go home early. Good work!

## Using this mindset at scale

There are tools available to help write code with such a mindset, that is to

> say writing code allowing all the valid and no impossible states.

Reviewer: ```suggestion
say writing code allowing only valid and no impossible states.
```

### Typings

Using a typed language of course helps a lot. Though they are not mandatory,
they certainly help developers avoid blind spots in their code. It's the
low-hanging fruit of code quality.

> When we writing this...

Reviewer: ```suggestion
When we write this...
```

```typescript
const someNumber: number = 2
```

...we're using the language to help remind ourself that valid states for
`someNumber` are all the numbers. Don't put a `string` in there.

> The challenge is to push the mindset the further we can, not just using what
> the language offer us (low-level types) but typing higher level code the same
> way. 

Reviewer: ```suggestion
The challenge is to push this simple idea as far as we can, not just using what
the language offers (low-level types) but typing higher level code using the same
basic principle. 
```

### Theory

Mathematics, type theory and functional programming also help. We should
especially be interested by [Alebraic Data
Type](https://en.wikipedia.org/wiki/Algebraic_data_type) (ADT) and more
precisely *sum types*.

Sum types are types where the value must be one of a *fixed* set of options.[^1]

> The `RemoteData` type we wrote is a sum types.

Reviewer: ```suggestion
The `RemoteData` type we wrote is a sum type.
```
Non?

Libraries that leverage those concepts and functional programming are strong
allies. The Typescript ecosystem has seen the addition of a lot of libraries in
that space.

> Don't bother writing the `RemoteData` by hand.
> [`RemoteData`](https://github.com/devexperts/remote-data-ts) does just that.
> [`fp-ts` also offer a lot of ADTs](https://github.com/gcanti/fp-ts) to model
> what's going on in our codebase like having *either* something or something else
> (Either) or having maybe something or not (Option).

Reviewer: ```suggestion
Don't bother writing the above `RemoteData` stuff by hand.
[`RemoteData`](https://github.com/devexperts/remote-data-ts) does just that.
[`fp-ts` also offer a lot of ADTs](https://github.com/gcanti/fp-ts) to model
what's going on in our codebase like having *either* something or something else
(Either) or maybe having something or not having it (Option).
```

The thing is, our applications are not that different. If you are faced with a
feature and ponders on its modelling as code, chances are a tool has been made
to help you do just that.

### Design

Last but not least, our brains! As software engineers, it's sometimes tempting

> to rush writing code without having analysed what we should represent, what are
> the valid states we should support.

Reviewer: ```suggestion
to rush into writing code without having analysed what we should represent and which valid states we should support.
```

It's not just about using technical and low-level or middle-level concepts such

> as `RemoteData`, `Option` or `Either`. Is first and foremost about using the
> same mindset to represent high level states the codebase should be in.

Reviewer: ```suggestion
as `RemoteData`, `Option` or `Either`. It is first and foremost about using the
right mindset to represent the high level states the codebase should be in.
```

You're writing a messaging app? Why not describe a message like this?

```typescript
type Message = EditingMessage | SendingMessage | SentMessage | FailedMessage

> ```

Reviewer: Je pense que tu peux t'en passer

Author: Okey je suppr !

---
title: The best test is the test you don't have to write

draft: false

page : false

image: /assets/img/best-test-trees.png

excerpt: >

>   Tests are heavily used to assert the quality of the codebase,
>   sometimes at the expense of investement in typed languages and advanced 
>   knowledge representation 

Reviewer: ```suggestion
  We heavily invest in tests to assert the quality of the codebase,
  but directing our efforts at typed languages and advanced 
  knowledge representation would make more sense.
```

---

![Abstract illustration of trees](/assets/img/best-test-trees.png)

When a software engineering team wants to improve or secure its codebase
quality, it often heavily invests in tests. 

> But tests are not a magic remedy. They're not free.

Reviewer: ```suggestion
But tests are not a magic remedy, and they are far from cheap.
```

> Using tests as the *only* tool that can help with quality can lead to more
> troubles. Are they worth the cost of their conception and maintenance? How can

Reviewer: ```suggestion
Using tests as the *only* tool that can help with quality can actually cause
troubles. Are they worth the cost of their conception and maintenance? How can
```

we maintain or improve the quality of a codebase without bloating it with
superficial tests?

## About tests

Let's say we wrote some code. Disciplined as we are, we know it's time to add
the corresponding test. We have the maturity to know we need different levels of
testing to cover a feature: unit testing, integration testing, end-to-end
testing, snapshot testing, etc.

It's a common reflex and our intentions are good. However, continuously adding
tests does create issues on its own. All the tests become a "codebase in the
codebase": with the dedicated libraries we use for them, the testing framework
that is so hard to change, the CI time required to run the whole test suite, the
maintenance those tests need, their flakiness, the colleagues complaining they

> need the best in class computer to run them in an acceptable time, etc.

Reviewer: ```suggestion
need the best-in-class computer to run them in an acceptable time, etc.
```

Tests are not free, so let's challenge our good intentions. Why are we writing
them in the first place?

We write tests to assert our code performs correctly under given constraints.
We know writing code is not easy, and a lot can go (and has gone) wrong.

## Being its own enemy

The key element is that tests are performed under "given contraints". If the
world always was in one single and predictable state, we would not have to write
and maintain any test: our code would run the way we've tested it by hand the
first time.

But the world can be in an infinite number of states. To be always safe, our
code and therefore our tests would need to accommodate as many of them as
possible.

It's especially true when we write code that will be used by other code. The
larger the API we write, the more states we allow and the more tests we will
have to write. Hell is paved with good intentions.

> The challenge here is for the code to cover all the use cases it needs to cover,
> but not more. Everything the code *can* do but *should not* do when used will
> require a test asserting it effectively isn't doing that.

Reviewer: ```suggestion
We don't need good intentions, we need properly-scoped ones. If we agree that everything our code *can* and *should not* do will require testing, then we should enforce strong limitations to our code's scope.
```

Impossible states are any states our code should not be in. A password that is
empty during a login. A payment that is at the same time waiting for approval
and rejected. A data we're fetching that is both loading and successful.

Let's take the example of the data we're fetching. We can call such data a
*remote* data. Imagine we're writing a naive React hook that does an
unspecified request and returns some data. One can use it like so:

```typescript
const {
  isLoading, // true | false 
  error,     // Error | undefined
  data,      // TheData | undefined
} = useMyRequest()
```

How many states can be represented by the API here? 2\*2\*2 = 8 states. Not
conviced? Here is the list of all of them:

<figure>
  <div markdown="1">

| isLoading | error | data |
|-----------|-----|------|
| true | undefined | undefined |
| true | undefined | Data |
| true | Error | undefined |
| true | Error | Data |
| false | undefined | undefined |
| false | undefined | Data |
| false | Error | undefined |
| false | Error | Data |

  </div>
  <figcaption>
    All the states we can represent
  </figcaption>
</figure>

That's a lot of possibilities and some of them are clearly not needed. After

> all, how many actual states can or should a naive remote data be in? Probably

Reviewer: ```suggestion
all, how many actual states can or should a naive remote data fetcher be in? Probably
```

only 4: request not yet asked, request loading, request failure, request
success.

The API we chose with the hook above therefore leaves us with 4 valid and 4
impossible states:

<figure>
  <div markdown="1">

| isLoading | error | data |
|-----------|-----|------|
| true | undefined | undefined |
| false | undefined | undefined |
| false | undefined | Data |
| false | Error | undefined |

  </div>
  <figcaption>
  The valid states
  </figcaption>
</figure>

`false | Error | undefined` for example means "request completed, but resulted
in failure", which is one of the valid states.

<figure>
  <div markdown="1">

| isLoading | error | data |
|-----------|-----|------|
| true | undefined | Data |
| true | Error | undefined |
| true | Error | Data |
| false | Error | Data |

  </div>
  <figcaption>
  The impossible states
  </figcaption>
</figure>

`true | Error | Data` means nothing for our naive request (both error and data
while still loading). It's an impossible state unrepresentative of the actual,
  actionable code functionality.

> To assert those 4 impossible states never happen despite the API allowing them,

Reviewer: ```suggestion
To ensure those 4 impossible states never happen despite the API allowing them,
```

we would need to write tests around the API, like an alarm system. "Hey! The
things you knew could happen but did not want to happen... they did happen".

Is it irremediable? It should not be. There is enough valid states to handle
already to not add impossible ones.

So, how to [make impossible states
impossible](https://www.youtube.com/watch?v=IcgmSRJHu_8)?

## Representing knowledge in code

[There is more to software engineering than continuously producing code (and
tests)](http://guillaume.wuips.com/tech/posts/2020-12-01-defining-success-engineering-team).
We sometimes have to pause and reflect on what models we're using, what business
features we're implementing. We can then use code to write this knowledge down.

> Now that we know our remote data can be in only 4 possibles states, how does it
> look like in code?

Reviewer: ```suggestion
Now that we know our remote data can only be in 4 possibles states, how do we enforce it with code rather than tests?
```

Such a mindset is something we're more used to see in codebases written with
ML-languages like Haskell:

```haskell
data RemoteData = NotAsked
                | Loading
                | Failure Error
                | Success TheData
```

But nothing prevents us from writing something similar with more familiar
languages, here in Typescript: 

```typescript
type RemoteData<E, P> = 
  | { type: 'NotAsked' }
  | { type: 'Loading' }
  | { type: 'Failure', error: E } 
  | { type: 'Success', payload: P }
```

This typing explicits our targeted benefits. The request can't be at the same

> time in loading and with an error. 

Reviewer: ```suggestion
time loading and with an error. 
```

Working with such a model is both easy and secure: we're forced to handle each
state in order to eventually access the one with data, and we're limited to the
4 states defined earlier:

```tsx
function Component() {
  const request = useMyRequest()

  switch (request._type) {
    case 'NotAsked':
      return <NotAskedScreen />

    case 'Loading':
      return <LoadingScreen />

    case 'Failure':
      return <ErrorScreen error={request.error} />

    case 'Success':
      return <SuccessScreen data={request.payload} />
  }
}
```

We have covered all valid states and no impossible ones. The developper writing
the `useMyRequest` hook has 4 less tests to write: the typing of the hook itself
enforces the output can ever only be these four cases. Testing it further would

> be testing the language itself. Developers using the hook also have 4 less tests

Reviewer: ```suggestion
be testing the language itself. We have scoped our code properly. Developers using the hook also have 4 less tests
```

to write as a result. Everyone is happy and can go home early. Good work!

## Using this mindset at scale

> There are tools available to help write code with such a mindset, that is to
> say writing code allowing only valid and no impossible states.

Reviewer: ```suggestion
There are tools available to help us write code focused on producing no impossible states.
```

### Typings

Using a typed language of course helps a lot. Though they are not mandatory,
they certainly help developers avoid blind spots in their code. It's the
low-hanging fruit of code quality.

When we write this...

```typescript
const someNumber: number = 2
```

...we're using the language to help remind ourself that valid states for
`someNumber` are all the numbers. Don't put a `string` in there.

The challenge is to push this simple idea as far as we can, not just using what
the language offers (low-level types) but typing higher level code using the

> same basic principle. 

Reviewer: ```suggestion
same basic principle. 

As we've seen, exploiting typed languages save on testing time. If the language itself makes it impossible for `2+2` to ever equals anything but `4`, why would we need to double check? What our code can and cannot do is no longer only our responsibility as developers, it is now also woven within the very fabric of it.
```
